src/app/modules/main/stickers/controller.nim

```diff
@@ -174,7 +192,31 @@ proc getStatusToken*(self: Controller): string =
   }
   return $jsonObj
 
-proc authenticateUser*(self: Controller, keyUid = "") =
+proc authenticate*(self: Controller, keyUid = "") =
   let data = SharedKeycarModuleAuthenticationArgs(uniqueIdentifier: UNIQUE_BUY_STICKER_TRANSACTION_MODULE_IDENTIFIER,
     keyUid: keyUid)
   self.events.emit(SIGNAL_SHARED_KEYCARD_MODULE_AUTHENTICATE_USER, data)
+
+proc disconnectKeycardReponseSignal(self: Controller) =
+  self.events.disconnect(self.connectionKeycardResponse)
+
+proc connectKeycardReponseSignal(self: Controller) =
+  self.connectionKeycardResponse = self.events.onWithUUID(SIGNAL_KEYCARD_RESPONSE) do(e: Args):
+    let args = KeycardLibArgs(e)
+    self.disconnectKeycardReponseSignal()
+    let currentFlow = self.keycardService.getCurrentFlow()
+    if currentFlow != KCSFlowType.Sign:
+      self.delegate.onTransactionSigned("", KeycardEvent())
+      return
+    self.delegate.onTransactionSigned(args.flowType, args.flowEvent)
+
+proc cancelCurrentFlow*(self: Controller) =
+  self.keycardService.cancelCurrentFlow()
+  # in most cases we're running another flow after canceling the current one,
+  # this way we're giving to the keycard some time to cancel the current flow
+  sleep(200)
```

Comment:
@IvanBelyakoff usually, when we're dealing with the keycard's calls, we rarely do just a single call, usually there is more than one call. If we don't stop here, letting keycard lib get ready for the next flow execution we we will block the keycard and won't be able to proceed.

Reply:
When we cancel the keycard flow, we have to give some minimal time to a keycard to get ready to accept the next flow, since it operates on a low level and needs much more time. This sleep, as mentioned above, is only to make a short pause between two or more successive calls. As far as I know, at this moment, there is no signal emitted from the keycard lib that would say the lib is ready for the next call.

As you said, switching between states is state matching, but while you're in some state to fulfill conditions to switch to some other state some keycard interaction must happen, often we have more than a single keycard flow run, one after another, so this sleep is related only to a running keycard flows from a single state and doesn't interfere to the rest of the app.

There are several places in the app where we're using the same approach, if you want to make them work without sleep, go for it, that's great, just make sure that all flows work correctly after that.